### PR TITLE
Fix Issue 23234 - Delegate literal with inferred return value that re…

### PR DESCRIPTION
**Pull request by RazvanN7**
…quires following alias-this uses class cast instead

### CODE REVIEW

**Comment by FeepingCreature** [comment]
Hm, why not `implicitCastTo`? Is the return type supposed to do an implicit hard cast here?

**Comment by maxhaton** [comment]
> Hm, why not `implicitCastTo`? Is the return type supposed to do an implicit hard cast here?

dmd seems to internally consider the existence of an alias this with the correct return type to be a way to implicitly cast

File: src/dmd/func.d [hunk 1]
```diff
@@ -3749,7 +3749,18 @@ extern (C++) final class FuncLiteralDeclaration : FuncDeclaration
...
                 if (exp && !exp.type.equals(tret))
                 {
-                    s.exp = exp.castTo(sc, tret);
+                    /* https://issues.dlang.org/show_bug.cgi?id=23234
+
+                    If the return statement of a FuncLiteralDeclaration
+                    is rewritten to match an expected type, we should
+                    check alias this before bluntly casting.
+                    */
+                    import dmd.aliasthis : resolveAliasThis;
```
Comment (maxhaton) [inline src/dmd/func.d:3758]: The change in behaviour needs to be documented at the top of modifyReturns too

Comment (maxhaton) [reply src/dmd/func.d:3758]: Also can anything fail at this point semantically?

Comment (FeepingCreature) [reply src/dmd/func.d:3758]: (To be clear, this is not so much a change in behavior as a miscompilation. The existing code silently creates crashes.)

**Comment by FeepingCreature** [comment]
> dmd seems to internally consider the existence of an alias this with the correct return type to be a way to implicitly cast

Right, so shouldn't `implicitCastTo` pick that up without requiring explicit invocation of alias-this?

File: src/dmd/func.d (hunk 1 above)
Comment (RazvanN7) [reply src/dmd/func.d:3758]: I think that the idea here was that errors should be avoided. `implicitCastTo` may issue some errors if that is not possible, whereas with the current implementation, the return expression is simply casted to the expected type. If it were me, I would just drop the callsite-based return type reference.

Comment (FeepingCreature) [reply src/dmd/func.d:3758]: I mean, but shouldn't the return type inference already be based on what sort of types can be implicitly converted?  For instance-
```
void foo(bool delegate() dg) { }
void main() { foo({ return 5; }); }
```
This errors, even though 5 can be `castTo` bool.

**Comment by FeepingCreature** [comment]
I've tried just doing an implicit cast, and it seems to pass the DMD testsuite at least
[just-do-an-implicit-cast.txt](https://github.com/dlang/dmd/files/9111128/just-do-an-implicit-cast.txt)
.

**Comment by RazvanN7** [comment]
Let's see what the testing pipeline thinks of this.

**Comment by RazvanN7** [comment]
@maxhaton @dkorpel @thewilsonator ready to get this in?

**Review by thewilsonator — APPROVED** [review]

File: compiler/test/runnable/test23234.d [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// https://issues.dlang.org/show_bug.cgi?id=23234
+
+class Bar
+{
+}
+
+class Foo
+{
+    Bar get() { return new Bar; }
```
Comment (maxhaton) [inline compiler/test/runnable/test23234.d:9]: Check that the alias this actually gets called?

**Review by ibuclaw — COMMENTED** [review]
Target master!

**Pull request merged**